docs/encyclopedia/application-message-passing.mdx

```diff
@@ -139,76 +197,161 @@ Stack Trace Queries are available only for running Workflow Executions.
+When Signals, Updates, and Queries arrive at your Workflow, the handlers for these messages will operate on the current state of your Workflow and can use whatever fields you have set.
+In this section, we’ll give you an overview of how messages work with Temporal and cover how to write correct and robust handlers by covering topics like atomicity, guaranteeing completion before the Workflow exits, exceptions, and idempotency.
 
-:::tip Support, stability, and dependency info
+### Message handler concurrency {#message-handler-concurrency}
 
-- In [Pre-release](/evaluate/release-stages#pre-release) (API is subject to change)
-- Introduced in [Temporal Server version 1.21](https://github.com/temporalio/temporal/releases/tag/v1.21.0)
-- Available in [Go SDK](https://pkg.go.dev/go.temporal.io/sdk@v1.23.1/client#Client.UpdateWorkflowWithOptions) since [v1.23.0](https://github.com/temporalio/sdk-go/releases/tag/v1.23.0)
-- Available in [Java SDK](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/client/WorkflowStub.html#startUpdate(io.temporal.client.UpdateOptions,java.lang.Object...)) since [v1.20.0](https://github.com/temporalio/sdk-java/releases/tag/v1.20.0)
-- Available in [Python SDK](https://docs.temporal.io/develop/python/features#updates) since [v1.4.0](https://github.com/temporalio/sdk-python/releases/tag/1.4.0)
-- Available in [.NET SDK](https://dotnet.temporal.io/api/Temporalio.Client.WorkflowHandle.html#Temporalio_Client_WorkflowHandle_ExecuteUpdateAsync_System_String_System_Collections_Generic_IReadOnlyCollection_System_Object__Temporalio_Client_WorkflowUpdateOptions_) since [v0.1.0-beta2](https://github.com/temporalio/sdk-dotnet/releases/tag/0.1.0-beta2)
-- Available in [TypeScript SDK](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle#executeupdate) since [v1.9.0](https://github.com/temporalio/sdk-typescript/releases/tag/v1.9.0)
-- Available in [PHP SDK](https://php.temporal.io/classes/Temporal-Client-WorkflowStubInterface.html#method_startUpdate) since [v2.8.0](https://github.com/temporalio/sdk-php/releases/tag/v2.8.0)
+If your Workflow receives messages, you may need to consider how those messages interact with one another or with the main Workflow method.
+Behind the scenes, Temporal is running a loop that looks like this:
+
+<PrettyImage src="/img/messages-workflow-loop.png" title="Diagram that shows the execution ordering of workflows" />
+
+Every time the Workflow wakes up--generally, it wakes up when it needs to--it will process messages in the order they were received, followed by making progress in the Workflow’s main method.
+
+This execution is on a single thread–while this means you don’t have to worry about parallelism, you do need to worry about concurrency if you have written Signal and Update handlers that can block. These can run interleaved with the main Workflow and with one another, resulting in potential race conditions. These methods should be made reentrant.
```

Comment:
[The event-loop diagram needs some updates along the lines of the changes I've proposed above.]

Reply:
I don't prefer all of these suggested changes, and I'm not sure what you're really going for with them.  I know you're trying to save me cycles, but it would be more helpful to get your experience reading than prescriptive non-trivial edits like this.

Reply:
(In particular, I fear you may be trying to do too much/be too comprehensive in this paragraph.)

Reply:
Will need more precise suggestions if you want the diagram changed.  Please follow up on slack.

Reply:
The main change I'm making here relates to

> Every time the Workflow wakes up [...] it will process messages in the order they were received, followed by making progress in the Workflow’s main method.

Our main aim in this section is to help users program safely in the face of interleaving, and my feeling is that the above description should be made a bit more accurate / explicit.  Specifically, it gives the impression that message processing happens, then the main workflow. But this downplays the opportunity for interleaving:
what's really happening is that a population of coroutines are set running: handlers first, until they reach their first yield points, then the main workflow coroutine. Thereafter, all coroutines have a chance to execute concurrently, with interleaving, until eventually no further progress can be made. The changes to the diagram would reflect the preceding sentences.

Yes, this is technical, but this is a technical section: we're introducing a concept of the "event loop" of Temporal workflows. I think these changes are not a great increase in technicality over the current content, and will help readers understand the safety challenges they face.

Reply:
Dan helped me update the diagram.